### PR TITLE
Custom CSS: change CSS link tag to inline style for performance

### LINKED ISSUE #1026: Idea to improve Custom CSS for huge and high traffic websites
I manage several sites with almost 1000 post and 2000 tags, very well positioned in Google, thus receiving 5k-7k daily visitors, and I noticed that server load skyrocketed from 2 to 5 in rush hours when using JetPack's Custom CSS. Even when I use W3TC to create html cache which is refreshed only once daily. But I'm noticing a lot of Apache calls to "?custom-css=1&csblog=1&cscache=6&csrev=10" which I traced back to Custom CSS:

By doing that, every visitor hit is bypassing the html cache triggering a db query, which is what we need to minimize.

Is there a possibility to change the way you save the CSS to use a physically existing -and cachable- file?
It will improve the amount of queries for big sites and constrained servers.

Of note: I use a couple themes which don't include settings for custom CSS and are receiving monthly updates. Also, those sites are built on the original themes, rather than child ones. That's why  we cannot simply update the CSS files, nor always able to recall these customization, since the only thing changed in the theme. I give support for WordPress owners and this is a very common issue for a lot of websites. And this proposed improvement could help a lot of people and servers to improve performance of their websites.

Hope to hear soon from you.
Thank you

### PR DESCRIPTION
**Pull request by sigginet**
The CSS `<link>` tag was changed to inline `<style>` tag for improved performance (fewer PHP requests) and cache'ability.
The URL param grabber (CSS code output code on top) was removed since it's no longer needed.

Fixes #1026

### CODE REVIEW

**Comment by georgestephanis** [comment]
Thanks for the pull, but I'm thinking we should really resolve this issue like

https://github.com/Automattic/jetpack/compare/custom-css-inline-it

-- I could have sworn I'd send a pull and merged that in, but it seems I'd lost track of it back in late August / early September.

@samhotchkiss -- if that looks good to you, can you generate and merge the pull for the `custom-css-inline-it` branch?

**Comment by sigginet** [comment]
Sounds cool .. I thought about adding a max filesize check, then thought 'nah .. later'.

To add to the discussion about when to inline vs external file, experts took a look at it in 2011 (see here) - https://mathiasbynens.be/notes/inline-vs-separate-file

But it gets trickier since WP uses PHP to deliver this CSS file so there's a MASSIVE overhead often (on shared servers or non-performance-optimized servers) - 503 ms in my client's case - http://www.webpagetest.org/result/150130_QV_16XP/1/details/#request10

By using a static file solution rather than PHP, you can reduce the response time down to ~150-200 ms - http://www.webpagetest.org/result/150130_QV_16XP/1/details/#request9

But here's the kicker though .. inlining these 4.5k (yes - more than 2k) would reduce the download time to 14ms ..

So, to sum things up for these 4.5kB:
503 ms - External PHP file on a shared/normal server
200 ms - External static CSS file
14  ms - Inline

I'm really struggling to see the reason behind insisting on external files for anything less than at LEAST 50k

I know .. I know, 50k sounds like a lot, but reality trumps theory ;)

**Comment by georgestephanis** [comment]
@sigginet The fact of the matter is that your client's 500ms lag really is out of the ordinary (have you considered a persistent object cache?) -- and caching the resulting css to a static file -- while it may work for some hosting setups, for others like large multisite setups, it could be positively disastrous, as each server behind the load balancer would wind up generating a css file and writing to disk for each hosted site -- which could be a lot.

The linked solution would be trivial for you to modify to always print it inline if you like -- a single line would do it:

```php
add_filter( 'safecss_embed_style', '__return_true' );
```

and that will do it.  I'll be doing the opposite for WPCOM -- adding `__return_false` -- to ensure that it never embeds inline, as per the conversation on that commit.

**Comment by sigginet** [comment]
@georgestephanis I know the 500ms lag is extreme. This is one of the reasons he hired me to optimize the website.
But this is how many private servers act out of the box, as well as shared hosting. (unfortunately)

How about having the pie and eating it too?
- In Multisite, have it configurable by the multisite admin (static file or not - default not static)
- In Multi & Single, inline when you're below X KB (I'd say 20kB since you should think about gzipped transfer, not uncompressed - gz will give you ~ 10x compression on CSS)
- In Singlesite, enable static file by default.
- In WPCOM, no inlining or static files (since WPCOM is optimized to deliver this quickly)

This way, it should be configured to deliver the optimal performance boost for ALL scenarios. Right?

Your thoughts?

**Comment by georgestephanis** [comment]
I'd be happy to add a hook to let you build that as a bolt on secondary plugin, but I'd prefer to avoid adding that much complexity to what should really be a simple thing in Jetpack core.

**Comment by georgestephanis** [comment]
Plus, static file generation -- if run when the admin saves the new css -- will not work in a load balanced single-site environment, unless there is some rsync setup in place -- which certainly can't be assumed.

**Comment by sigginet** [comment]
@georgestephanis true .. but how does a load balanced environment then handle uploaded images, etc?
It must handle static files in some manner .. the static CSS files should be handled in the same manner .. (the static file would be created in something like /wp-content/uploads/ or /wp-content/cache/ )

Aaaanywho, I will be using CSS inlining in this client's case.

Thanks for taking a look :)

**Comment by georgestephanis** [comment]
#1592 is the PR for my take, will get it merged after the pending 3.4 release.

**Comment by georgestephanis** [comment]
Fixed by #1592
